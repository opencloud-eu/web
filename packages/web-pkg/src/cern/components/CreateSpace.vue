<template>
  <oc-button
    id="new-space-menu-btn"
    v-oc-tooltip="$gettext('Request new project')"
    type="a"
    href="https://cern.service-now.com/service-portal?id=sc_cat_item&name=request-storage-space&se=CERNBox-Service"
    target="_blank"
    appearance="filled"
  >
    <oc-icon name="add" />
    <span v-translate>New project</span>
  </oc-button>
</template>
