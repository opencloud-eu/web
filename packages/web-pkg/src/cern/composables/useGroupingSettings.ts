import { IncomingShareResource } from '@opencloud-eu/web-client'
import { computed, Ref, unref } from 'vue'

export const useGroupingSettings = ({
  sortBy,
  sortDir
}: {
  sortBy: Ref<string>
  sortDir: Ref<string>
}) => {
  const groupingSettings = computed(() => {
    return {
      groupingBy: localStorage.getItem('grouping-shared-with-me') || 'Shared on',
      showGroupingOptions: true,
      groupingFunctions: {
        'Name alphabetically': function (row: IncomingShareResource) {
          localStorage.setItem('grouping-shared-with-me', 'Name alphabetically')
          if (!isNaN(Number(row.name.charAt(0)))) {
            return '#'
          }
          if (row.name.charAt(0) === '.') {
            return row.name.charAt(1).toLowerCase()
          }
          return row.name.charAt(0).toLowerCase()
        },
        'Shared on': function (row: IncomingShareResource) {
          localStorage.setItem('grouping-shared-with-me', 'Shared on')
          const recently = Date.now() - 604800000
          const lastMonth = Date.now() - 2592000000
          if (Date.parse(row.sdate) < lastMonth) {
            return 'Older'
          }
          if (Date.parse(row.sdate) >= recently) {
            return 'Recently'
          } else {
            return 'Last month'
          }
        },
        'Share owner': function (row: IncomingShareResource) {
          localStorage.setItem('grouping-shared-with-me', 'Share owner')
          return row?.owner?.displayName
        },
        None: function () {
          localStorage.setItem('grouping-shared-with-me', 'None')
        }
      },
      sortGroups: {
        'Name alphabetically': function (groups: { name: string }[]) {
          // sort in alphabetical order by group name
          const sortedGroups = groups.sort(function (a, b) {
            if (a.name < b.name) {
              return -1
            }
            if (a.name > b.name) {
              return 1
            }
            return 0
          })
          // if sorting is done by name, reverse groups depending on asc/desc
          if (unref(sortBy) === 'name' && unref(sortDir) === 'desc') {
            sortedGroups.reverse()
          }
          return sortedGroups
        },
        'Shared on': function (groups: { name: string }[]) {
          // sort in order: 1-Recently, 2-Last month, 3-Older
          const sortedGroups = []
          const options = ['Recently', 'Last month', 'Older']
          for (const o of options) {
            const found = groups.find((el) => el.name.toLowerCase() === o.toLowerCase())
            if (found) {
              sortedGroups.push(found)
            }
          }
          // if sorting is done by sdate, reverse groups depending on asc/desc
          if (unref(sortBy) === 'sdate' && unref(sortDir) === 'asc') {
            sortedGroups.reverse()
          }
          return sortedGroups
        },
        'Share owner': function (groups: { name: string }[]) {
          // sort in alphabetical order by group name
          const sortedGroups = groups.sort(function (a, b) {
            if (a.name < b.name) {
              return -1
            }
            if (a.name > b.name) {
              return 1
            }
            return 0
          })
          // if sorting is done by owner, reverse groups depending on asc/desc
          if (unref(sortBy) === 'owner' && unref(sortDir) === 'desc') {
            sortedGroups.reverse()
          }
          return sortedGroups
        }
      }
    }
  })

  return {
    groupingSettings
  }
}
