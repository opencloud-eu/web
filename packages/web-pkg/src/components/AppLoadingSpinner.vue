<template>
  <div class="flex items-center justify-center oc-height-1-1 w-full">
    <oc-spinner id="app-loading-spinner" size="large" :aria-hidden="true" aria-label="" />
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'AppLoadingSpinner'
})
</script>
