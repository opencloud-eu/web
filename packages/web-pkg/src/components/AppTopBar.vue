<template>
  <portal to="app.runtime.header.left">
    <div class="oc-app-top-bar oc-flex">
      <span class="oc-app-top-bar-inner oc-pl-m oc-pr-xs oc-flex oc-flex-middle oc-flex-between">
        <div class="open-file-bar oc-flex">
          <resource-list-item
            v-if="resource"
            id="app-top-bar-resource"
            :is-thumbnail-displayed="false"
            :is-extension-displayed="areFileExtensionsShown"
            :path-prefix="getPathPrefix(resource)"
            :resource="resource"
            :parent-folder-name="getParentFolderName(resource)"
            :parent-folder-link-icon-additional-attributes="
              getParentFolderLinkIconAdditionalAttributes(resource)
            "
            :is-path-displayed="isPathDisplayed"
            :is-resource-clickable="false"
          />
        </div>
        <div class="oc-flex main-actions">
          <template v-if="dropDownMenuSections.length">
            <oc-button
              id="oc-openfile-contextmenu-trigger"
              v-oc-tooltip="contextMenuLabel"
              :aria-label="contextMenuLabel"
              appearance="raw-inverse"
              color-role="chrome"
              class="oc-p-xs"
            >
              <oc-icon name="more-2" />
            </oc-button>
            <oc-drop
              drop-id="oc-openfile-contextmenu"
              mode="click"
              padding-size="small"
              toggle="#oc-openfile-contextmenu-trigger"
              close-on-click
              :title="resource.name"
              @click.stop.prevent
            >
              <context-action-menu
                :menu-sections="dropDownMenuSections"
                :action-options="dropDownActionOptions"
              />
            </oc-drop>
          </template>
          <span v-if="hasAutosave" class="oc-flex oc-flex-middle">
            <oc-icon
              v-oc-tooltip="autoSaveTooltipText"
              :accessible-label="autoSaveTooltipText"
              name="refresh"
              color="white"
              class="ox-p-xs oc-mx-xs"
            />
          </span>
          <template v-if="mainActions.length && resource">
            <context-action-menu
              :menu-sections="[
                {
                  name: 'main-actions',
                  items: mainActions
                    .filter((action) => action.isVisible())
                    .map((action) => {
                      return {
                        ...action,
                        class: 'oc-p-xs app-topbar-action',
                        hideLabel: true
                      }
                    })
                }
              ]"
              :action-options="{
                resources: [resource]
              }"
              appearance="raw-inverse"
              color-role="chrome"
            />
          </template>
          <oc-button
            id="app-top-bar-close"
            appearance="raw-inverse"
            color-role="chrome"
            class="oc-p-xs"
            :aria-label="$gettext('Close')"
            @click="$emit('close')"
          >
            <oc-icon name="close" />
          </oc-button>
        </div>
      </span>
    </div>
  </portal>
</template>

<script lang="ts">
import { computed, defineComponent, PropType, unref } from 'vue'
import ContextActionMenu from './ContextActions/ContextActionMenu.vue'
import { useGettext } from 'vue3-gettext'
import {
  Action,
  FileActionOptions,
  useConfigStore,
  useFolderLink,
  useGetMatchingSpace,
  useResourcesStore
} from '../composables'
import ResourceListItem from './FilesList/ResourceListItem.vue'
import { isPublicSpaceResource, Resource } from '@opencloud-eu/web-client'
import { Duration } from 'luxon'
import { MenuSection } from './ContextActions'

export default defineComponent({
  name: 'AppTopBar',
  components: {
    ContextActionMenu,
    ResourceListItem
  },
  props: {
    dropDownMenuSections: {
      type: Array as PropType<MenuSection[]>,
      default: (): MenuSection[] => []
    },
    dropDownActionOptions: {
      type: Object as PropType<FileActionOptions>,
      default: (): FileActionOptions => ({
        space: null,
        resources: []
      })
    },
    mainActions: {
      type: Array as PropType<Action[]>,
      default: (): Action[] => []
    },
    hasAutoSave: {
      type: Boolean,
      default: true
    },
    isEditor: {
      type: Boolean,
      default: false
    },
    resource: {
      type: Object as PropType<Resource>,
      default: null
    }
  },
  emits: ['close'],
  setup(props) {
    const { $gettext, current: currentLanguage } = useGettext()
    const resourcesStore = useResourcesStore()
    const configStore = useConfigStore()
    const { getMatchingSpace } = useGetMatchingSpace()

    const areFileExtensionsShown = computed(() => resourcesStore.areFileExtensionsShown)
    const contextMenuLabel = computed(() => $gettext('Show context menu'))
    const hasAutosave = computed(
      () => props.isEditor && props.hasAutoSave && configStore.options.editor.autosaveEnabled
    )
    const autoSaveTooltipText = computed(() => {
      const duration = Duration.fromObject(
        { seconds: configStore.options.editor.autosaveInterval },
        { locale: currentLanguage }
      )
      return $gettext(`Autosave (every %{ duration })`, { duration: duration.toHuman() })
    })

    const space = computed(() => getMatchingSpace(props.resource))

    const isPathDisplayed = computed(() => {
      return !isPublicSpaceResource(unref(space))
    })

    return {
      contextMenuLabel,
      areFileExtensionsShown,
      hasAutosave,
      autoSaveTooltipText,
      isPathDisplayed,
      ...useFolderLink()
    }
  }
})
</script>

<style lang="scss">
.oc-app-top-bar {
  align-self: center;
  grid-column: 1 / 4;
  grid-row: secondRow;

  @media (min-width: $oc-breakpoint-small-default) {
    grid-column: 2;
    grid-row: 1;
  }
}

.oc-app-top-bar-inner {
  align-self: center;
  background-color: var(--oc-role-chrome);
  border-radius: 10px;
  border: 1px solid var(--oc-role-on-chrome);
  display: inline-flex;
  gap: 25px;
  height: 40px;
  margin: 10px auto;
  width: 100%;

  @media (min-width: $oc-breakpoint-small-default) {
    flex-basis: 250px;
    margin: 0;
  }

  .oc-resource-indicators {
    .text {
      color: var(--oc-role-on-chrome);
    }
  }
}

.app-topbar-action {
  color: var(--oc-role-on-chrome) !important;

  svg {
    fill: var(--oc-role-on-chrome) !important;
  }

  &:hover:not(:disabled) {
    color: var(--oc-role-on-surface) !important;

    svg {
      fill: var(--oc-role-on-surface) !important;
    }
  }
}

.open-file-bar {
  #app-top-bar-resource {
    max-width: 360px;

    @media (max-width: $oc-breakpoint-medium-default) {
      max-width: 240px;
    }

    @media (min-width: $oc-breakpoint-small-default) {
      widows: initial;
    }

    svg,
    .oc-resource-name span {
      fill: var(--oc-role-on-chrome) !important;
      color: var(--oc-role-on-chrome) !important;
    }
  }

  .oc-resource-icon:hover,
  .oc-resource-name:hover {
    cursor: default;
    text-decoration: none;
  }
}
</style>
