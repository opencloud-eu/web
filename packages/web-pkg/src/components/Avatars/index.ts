export { default as UserAvatar } from './UserAvatar.vue'
export { default as AvatarUpload } from './AvatarUpload.vue'
