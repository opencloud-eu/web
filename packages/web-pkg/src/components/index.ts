export * from './AppBar'
export * from './AppTemplates'
export * from './ContextActions'
export * from './FilesList'
export * from './Filters'
export * from './Modals'
export * from './SideBar'
export * from './Search'
export * from './Spaces'
export * from './TextEditor'

export { default as AppLoadingSpinner } from './AppLoadingSpinner.vue'
export { default as AppTopBar } from './AppTopBar.vue'
export { default as BatchActions } from './BatchActions.vue'
export { default as CustomComponentTarget } from './CustomComponentTarget.vue'
export { default as ItemFilter } from './ItemFilter.vue'
export { default as ItemFilterToggle } from './ItemFilterToggle.vue'
export { default as LinkRoleDropdown } from './LinkRoleDropdown.vue'
export { default as LoadingIndicator } from './LoadingIndicator.vue'
export { default as NoContentMessage } from './NoContentMessage.vue'
export { default as Pagination } from './Pagination.vue'
export { default as QuotaSelect } from './QuotaSelect.vue'
export { default as SpaceQuota } from './SpaceQuota.vue'
export { default as SearchBarFilter } from './SearchBarFilter.vue'
export { default as ViewOptions } from './ViewOptions.vue'
export { default as PortalTarget } from './PortalTarget.vue'
export { default as CreateShortcutModal } from './CreateShortcutModal.vue'
export { default as CreateLinkModal } from './CreateLinkModal.vue'
