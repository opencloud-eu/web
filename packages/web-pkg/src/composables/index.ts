export * from './ability'
export * from './actions'
export * from './appDefaults'
export * from './appProviderService'
export * from './archiverService'
export * from './authContext'
export * from './breadcrumbs'
export * from './clientService'
export * from './clipboard'
export * from './download'
export * from './driveResolver'
export * from './embedMode'
export * from './eventBus'
export * from './fileListHeaderPosition'
export * from './filesList'
export * from './folderLink'
export * from './isTopBarSticky'
export * from './keyboardActions'
export * from './links'
export * from './loadingService'
export * from './localStorage'
export * from './pagination'
export * from './passwordPolicyService'
export * from './piniaStores'
export * from './portalTarget'
export * from './previewService'
export * from './requestHeaders'
export * from './resources'
export * from './router'
export * from './scrollTo'
export * from './search'
export * from './selection'
export * from './service'
export * from './shares'
export * from './sideBar'
export * from './sort'
export * from './spaces'
export * from './upload'
export * from './viewMode'
export * from './webWorkers'
export * from './avatars'
