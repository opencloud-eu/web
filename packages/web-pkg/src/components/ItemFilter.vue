<template>
  <div class="item-filter flex" :class="`item-filter-${filterName}`">
    <oc-filter-chip
      :filter-label="filterLabel"
      :selected-item-names="selectedItems.map((i) => i[displayNameAttribute])"
      :close-on-click="closeOnClick"
      @clear-filter="clearFilter"
      @show-drop="showDrop"
    >
      <template #default>
        <oc-text-input
          v-if="showOptionFilter && filterableAttributes.length"
          ref="filterInputRef"
          v-model="filterTerm"
          class="item-filter-input mb-4 mt-2"
          autocomplete="off"
          :label="optionFilterLabel === '' ? $gettext('Filter list') : optionFilterLabel"
        />
        <div ref="itemFilterListRef">
          <oc-list class="item-filter-list">
            <li v-for="(item, index) in displayedItems" :key="index" class="my-1">
              <oc-button
                class="item-filter-list-item flex items-center w-full"
                :class="{
                  'item-filter-list-item-active': !allowMultiple && isItemSelected(item),
                  'justify-start': allowMultiple,
                  'justify-between': !allowMultiple
                }"
                justify-content="space-between"
                appearance="raw"
                :data-test-value="item[displayNameAttribute as keyof Item]"
                @click="toggleItemSelection(item)"
              >
                <div class="flex items-center truncate">
                  <oc-checkbox
                    v-if="allowMultiple"
                    size="large"
                    class="mr-2"
                    :label="$gettext('Toggle selection')"
                    :model-value="isItemSelected(item)"
                    :label-hidden="true"
                    @update:model-value="toggleItemSelection(item)"
                    @click.stop
                  />
                  <div>
                    <slot name="image" :item="item" />
                  </div>
                  <div class="truncate">
                    <slot name="item" :item="item" />
                  </div>
                </div>
                <div class="flex">
                  <oc-icon v-if="!allowMultiple && isItemSelected(item)" name="check" />
                </div>
              </oc-button>
            </li>
          </oc-list>
        </div>
      </template>
    </oc-filter-chip>
  </div>
</template>

<script lang="ts">
import {
  ComponentPublicInstance,
  PropType,
  defineComponent,
  nextTick,
  onMounted,
  ref,
  unref,
  useTemplateRef,
  watch
} from 'vue'
import Fuse, { FuseOptionKey } from 'fuse.js'
import Mark from 'mark.js'
import omit from 'lodash-es/omit'
import { useRoute, useRouteQuery, useRouter } from '../composables'
import { defaultFuseOptions } from '../helpers'
import { queryItemAsString } from '../composables/appDefaults'
import { OcTextInput } from '@opencloud-eu/design-system/components'

type Item = unknown

export default defineComponent({
  name: 'ItemFilter',
  props: {
    filterLabel: {
      type: String,
      required: true
    },
    filterName: {
      type: String,
      required: true
    },
    optionFilterLabel: {
      type: String,
      required: false,
      default: ''
    },
    showOptionFilter: {
      type: Boolean,
      required: false,
      default: false
    },
    items: {
      type: Array as PropType<Item[]>,
      required: true
    },
    allowMultiple: {
      type: Boolean,
      required: false,
      default: false
    },
    idAttribute: {
      type: String,
      required: false,
      default: 'id'
    },
    displayNameAttribute: {
      type: String,
      required: false,
      default: 'name'
    },
    filterableAttributes: {
      type: Array as PropType<FuseOptionKey<Item>[]>,
      required: false,
      default: (): FuseOptionKey<Item>[] => []
    },
    closeOnClick: {
      type: Boolean,
      default: false
    }
  },
  emits: ['selectionChange'],
  setup: function (props, { emit, expose }) {
    const router = useRouter()
    const currentRoute = useRoute()
    const filterInputRef =
      useTemplateRef<ComponentPublicInstance<typeof OcTextInput>>('filterInputRef')
    const selectedItems = ref<Item[]>([])
    const displayedItems = ref(props.items)
    const itemFilterListRef = useTemplateRef('itemFilterListRef')

    const queryParam = `q_${props.filterName}`
    const currentRouteQuery = useRouteQuery(queryParam)

    const getId = (item: Item) => {
      return item[props.idAttribute as keyof Item]
    }

    const setRouteQuery = () => {
      return router.push({
        query: {
          ...omit(unref(currentRoute).query, [queryParam]),
          ...(!!unref(selectedItems).length && {
            [queryParam]: unref(selectedItems)
              .reduce<string>((acc, item) => {
                acc += `${getId(item)}+`
                return acc
              }, '')
              .slice(0, -1)
          })
        }
      })
    }

    const isItemSelected = (item: Item) => {
      return !!unref(selectedItems).find((s) => getId(s) === getId(item))
    }

    const toggleItemSelection = async (item: Item) => {
      if (isItemSelected(item)) {
        selectedItems.value = unref(selectedItems).filter((s) => getId(s) !== getId(item))
      } else {
        if (!props.allowMultiple) {
          selectedItems.value = []
        }
        selectedItems.value.push(item)
      }
      await setRouteQuery()
      emit('selectionChange', unref(selectedItems))
    }

    const filterTerm = ref<string>()
    const filter = (items: Item[], filterTerm: string) => {
      if (!(filterTerm || '').trim()) {
        return items
      }
      const fuse = new Fuse(items, {
        ...defaultFuseOptions,
        keys: props.filterableAttributes
      })

      const results = fuse.search(filterTerm).map((r) => r.item)
      return items.filter((item) => results.includes(item))
    }
    const clearFilter = () => {
      selectedItems.value = []
      emit('selectionChange', unref(selectedItems))
      setRouteQuery()
    }

    const setDisplayedItems = (items: Item[]) => {
      displayedItems.value = items
    }

    const showDrop = async () => {
      setDisplayedItems(props.items)
      await nextTick()
      unref(filterInputRef)?.focus()
    }

    let markInstance: Mark | undefined
    watch(filterTerm, () => {
      setDisplayedItems(filter(props.items, unref(filterTerm)))
      if (unref(itemFilterListRef)) {
        markInstance = new Mark(unref(itemFilterListRef))
        markInstance.unmark()
        markInstance.mark(unref(filterTerm), {
          element: 'span',
          className: 'mark-highlight'
        })
      }
    })

    const setSelectedItemsBasedOnQuery = () => {
      const queryStr = queryItemAsString(unref(currentRouteQuery))
      if (queryStr) {
        const ids = queryStr.split('+')
        selectedItems.value = props.items.filter((s) => ids.includes(getId(s)))
      }
    }

    expose({ setSelectedItemsBasedOnQuery })

    onMounted(() => {
      setSelectedItemsBasedOnQuery()
    })

    return {
      clearFilter,
      displayedItems,
      filterInputRef,
      filterTerm,
      isItemSelected,
      itemFilterListRef,
      queryParam,
      selectedItems,
      setDisplayedItems,
      showDrop,
      toggleItemSelection,
      // expose to type
      setSelectedItemsBasedOnQuery
    }
  }
})
</script>
