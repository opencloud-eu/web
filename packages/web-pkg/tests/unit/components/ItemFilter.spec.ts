import ItemFilter from '../../../src/components/ItemFilter.vue'
import {
  PartialComponentProps,
  defaultComponentMocks,
  defaultPlugins,
  mount
} from '@opencloud-eu/web-test-helpers'
import { queryItemAsString } from '../../../src/composables/appDefaults'
import { OcCheckbox } from '@opencloud-eu/design-system/components'

vi.mock('../../../src/composables/appDefaults')

const filterItems = [
  { id: '1', name: 'Albert Einstein' },
  { id: '2', name: 'Marie Curie' }
]

const selectors = {
  filterInput: '.item-filter-input input',
  checkboxStub: 'oc-checkbox-stub',
  filterListItem: '.item-filter-list-item',
  activeFilterListItem: '.item-filter-list-item-active',
  clearBtn: '.oc-filter-chip-clear'
}

describe('ItemFilter', () => {
  it('renders all items', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('can use a custom attribute as display name', () => {
    const filterItems = [
      { id: '1', displayName: 'Albert Einstein' },
      { id: '2', displayName: 'Marie Curie' }
    ]
    const { wrapper } = getWrapper({
      props: { displayNameAttribute: 'displayName', items: filterItems }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
  describe('filter', () => {
    it('renders the input field when enabled', () => {
      const { wrapper } = getWrapper({
        props: { showOptionFilter: true, filterableAttributes: ['name'] }
      })
      expect(wrapper.find(selectors.filterInput).exists()).toBeTruthy()
    })
    it.each([
      { filterTerm: '', expectedResult: 2 },
      { filterTerm: 'Albert', expectedResult: 1 },
      { filterTerm: 'invalid', expectedResult: 0 }
    ])('filters on input', async (data) => {
      const { wrapper } = getWrapper({
        props: { showOptionFilter: true, filterableAttributes: ['name'] }
      })
      await wrapper.find(selectors.filterInput).setValue(data.filterTerm)
      expect(wrapper.findAll(selectors.filterListItem).length).toBe(data.expectedResult)
    })
  })
  describe('selection', () => {
    it('allows selection of multiple items on click', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: true } })
      expect(wrapper.emitted('selectionChange')).toBeFalsy()
      let selectionChangeEmits = 0
      for (const item of wrapper.findAll(selectors.filterListItem)) {
        expect(
          item.findComponent<typeof OcCheckbox>(selectors.checkboxStub).props('modelValue')
        ).toBeFalsy()
        await item.trigger('click')
        selectionChangeEmits += 1
        expect(wrapper.emitted('selectionChange').length).toBe(selectionChangeEmits)
        expect(
          item.findComponent<typeof OcCheckbox>(selectors.checkboxStub).props('modelValue')
        ).toBeTruthy()
      }
      expect(wrapper.vm.selectedItems.length).toBe(wrapper.findAll(selectors.filterListItem).length)
    })
    it('does not allow selection of multiple items when disabled', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: false } })
      const first = wrapper.findAll(selectors.filterListItem).at(0)
      await first.trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(1)
      expect(wrapper.find(selectors.activeFilterListItem).exists()).toBeTruthy()

      const second = wrapper.findAll(selectors.filterListItem).at(1)
      await second.trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(2)
      expect(wrapper.vm.selectedItems.length).toBe(1)
    })
    it('does de-select the current selected item on click', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: true } })
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      await item.trigger('click')
      await item.trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(2)
      expect(
        item.findComponent<typeof OcCheckbox>(selectors.checkboxStub).props('modelValue')
      ).toBeFalsy()
      expect(wrapper.vm.selectedItems.length).toBe(0)
    })
    it('clears the selection when the clear-button is being clicked', async () => {
      const { wrapper } = getWrapper({ props: { allowMultiple: true } })
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      await item.trigger('click')
      await wrapper.find(selectors.clearBtn).trigger('click')
      expect(wrapper.emitted('selectionChange').length).toBe(2)
      expect(
        item.findComponent<typeof OcCheckbox>(selectors.checkboxStub).props('modelValue')
      ).toBeFalsy()
      expect(wrapper.vm.selectedItems.length).toBe(0)
    })
  })
  describe('route query', () => {
    it('sets the selected item as route query param', async () => {
      const { wrapper, mocks } = getWrapper()
      const item = wrapper.findAll(selectors.filterListItem).at(0)
      expect(mocks.$router.push).not.toHaveBeenCalled()
      await item.trigger('click')
      expect(mocks.$router.push).toHaveBeenCalledWith(
        expect.objectContaining({
          query: expect.objectContaining({ [wrapper.vm.queryParam]: '1' })
        })
      )
    })
    it('sets the selected items initially when given via query param', () => {
      const { wrapper } = getWrapper({ initialQuery: '1' })
      expect(wrapper.vm.selectedItems).toEqual([filterItems[0]])
    })
  })

  describe('label prop', () => {
    it('sets the correct label using getLabel computed property', () => {
      const label = 'Filter groups'
      const { wrapper } = getWrapper({
        props: {
          showOptionFilter: true,
          filterableAttributes: ['name'],
          optionFilterLabel: label
        }
      })
      expect(wrapper.find('.item-filter-input label').text()).toBe(label)
    })

    it('sets the default label using getLabel computed property when no prop is set', () => {
      const label: string = undefined
      const { wrapper } = getWrapper({
        props: {
          showOptionFilter: true,
          filterableAttributes: ['name'],
          optionFilterLabel: label
        }
      })
      expect(wrapper.find('.item-filter-input label').text()).toBe('Filter list')
    })
  })
})

function getWrapper({
  props = {},
  initialQuery = ''
}: { props?: PartialComponentProps<typeof ItemFilter>; initialQuery?: string } = {}) {
  vi.mocked(queryItemAsString).mockImplementation(() => initialQuery)
  const mocks = defaultComponentMocks()
  return {
    mocks,
    wrapper: mount(ItemFilter, {
      props: {
        filterLabel: 'Users',
        filterName: 'users',
        items: filterItems,
        ...props
      },
      slots: {
        item(data) {
          return props.displayNameAttribute ? data.item[props.displayNameAttribute] : data.item.name
        }
      },
      global: {
        plugins: [...defaultPlugins()],
        mocks,
        provide: mocks,
        stubs: { OcCheckbox: true }
      }
    })
  }
}
