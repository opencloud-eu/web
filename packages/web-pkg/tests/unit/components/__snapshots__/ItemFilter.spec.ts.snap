// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemFilter > can use a custom attribute as display name 1`] = `
"<div class="item-filter oc-flex item-filter-users">
  <div class="oc-filter-chip oc-flex"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-surface oc-button-raw-inverse oc-button-surface-raw-inverse oc-filter-chip-button oc-pill" id="oc-filter-chip-3">
      <!--v-if-->
      <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-filter-check-icon-inactive"><svg data-testid="inline-svg-stub" src="icons/check-fill.svg" transform-source="(svg) => {
      if (__props.accessibleLabel !== &quot;&quot;) {
        const title = document.createElement(&quot;title&quot;);
        title.setAttribute(&quot;id&quot;, svgTitleId.value);
        title.appendChild(document.createTextNode(__props.accessibleLabel));
        svg.insertBefore(title, svg.firstChild);
      };
      return svg;
    }" aria-hidden="true" focusable="false"></svg></span> <span class="oc-text-truncate oc-filter-chip-label">Users</span>
      <!--v-if--> <span class="oc-icon oc-icon-s"><svg data-testid="inline-svg-stub" src="icons/arrow-down-s-fill.svg" transform-source="(svg) => {
      if (__props.accessibleLabel !== &quot;&quot;) {
        const title = document.createElement(&quot;title&quot;);
        title.setAttribute(&quot;id&quot;, svgTitleId.value);
        title.appendChild(document.createTextNode(__props.accessibleLabel));
        svg.insertBefore(title, svg.firstChild);
      };
      return svg;
    }" aria-hidden="true" focusable="false"></svg></span>
    </button>
    <div id="oc-drop-4" class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop">
      <div class="oc-card oc-card-body oc-p-s">
        <!--v-if-->
        <div>
          <ul class="oc-list oc-my-rm oc-mx-rm item-filter-list">
            <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw item-filter-list-item oc-flex oc-flex-middle oc-width-1-1 oc-p-s oc-flex-between" data-test-value="Albert Einstein">
                <!--v-if-->
                <!-- @slot Content of the button -->
                <div class="oc-flex oc-flex-middle oc-text-truncate">
                  <!--v-if-->
                  <div></div>
                  <div class="oc-text-truncate oc-ml-s">Albert Einstein</div>
                </div>
                <div class="oc-flex">
                  <!--v-if-->
                </div>
              </button></li>
            <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw item-filter-list-item oc-flex oc-flex-middle oc-width-1-1 oc-p-s oc-flex-between" data-test-value="Marie Curie">
                <!--v-if-->
                <!-- @slot Content of the button -->
                <div class="oc-flex oc-flex-middle oc-text-truncate">
                  <!--v-if-->
                  <div></div>
                  <div class="oc-text-truncate oc-ml-s">Marie Curie</div>
                </div>
                <div class="oc-flex">
                  <!--v-if-->
                </div>
              </button></li>
          </ul>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`ItemFilter > renders all items 1`] = `
"<div class="item-filter oc-flex item-filter-users">
  <div class="oc-filter-chip oc-flex"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-surface oc-button-raw-inverse oc-button-surface-raw-inverse oc-filter-chip-button oc-pill" id="oc-filter-chip-1">
      <!--v-if-->
      <!-- @slot Content of the button --> <span class="oc-icon oc-icon-s oc-filter-check-icon-inactive"><svg data-testid="inline-svg-stub" src="icons/check-fill.svg" transform-source="(svg) => {
      if (__props.accessibleLabel !== &quot;&quot;) {
        const title = document.createElement(&quot;title&quot;);
        title.setAttribute(&quot;id&quot;, svgTitleId.value);
        title.appendChild(document.createTextNode(__props.accessibleLabel));
        svg.insertBefore(title, svg.firstChild);
      };
      return svg;
    }" aria-hidden="true" focusable="false"></svg></span> <span class="oc-text-truncate oc-filter-chip-label">Users</span>
      <!--v-if--> <span class="oc-icon oc-icon-s"><svg data-testid="inline-svg-stub" src="icons/arrow-down-s-fill.svg" transform-source="(svg) => {
      if (__props.accessibleLabel !== &quot;&quot;) {
        const title = document.createElement(&quot;title&quot;);
        title.setAttribute(&quot;id&quot;, svgTitleId.value);
        title.appendChild(document.createTextNode(__props.accessibleLabel));
        svg.insertBefore(title, svg.firstChild);
      };
      return svg;
    }" aria-hidden="true" focusable="false"></svg></span>
    </button>
    <div id="oc-drop-2" class="oc-drop oc-box-shadow-medium oc-rounded oc-filter-chip-drop">
      <div class="oc-card oc-card-body oc-p-s">
        <!--v-if-->
        <div>
          <ul class="oc-list oc-my-rm oc-mx-rm item-filter-list">
            <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw item-filter-list-item oc-flex oc-flex-middle oc-width-1-1 oc-p-s oc-flex-between" data-test-value="Albert Einstein">
                <!--v-if-->
                <!-- @slot Content of the button -->
                <div class="oc-flex oc-flex-middle oc-text-truncate">
                  <!--v-if-->
                  <div></div>
                  <div class="oc-text-truncate oc-ml-s">Albert Einstein</div>
                </div>
                <div class="oc-flex">
                  <!--v-if-->
                </div>
              </button></li>
            <li class="oc-my-xs"><button type="button" class="oc-button oc-rounded oc-button-m oc-button-justify-content-space-between oc-button-gap-m oc-button-secondary oc-button-raw oc-button-secondary-raw item-filter-list-item oc-flex oc-flex-middle oc-width-1-1 oc-p-s oc-flex-between" data-test-value="Marie Curie">
                <!--v-if-->
                <!-- @slot Content of the button -->
                <div class="oc-flex oc-flex-middle oc-text-truncate">
                  <!--v-if-->
                  <div></div>
                  <div class="oc-text-truncate oc-ml-s">Marie Curie</div>
                </div>
                <div class="oc-flex">
                  <!--v-if-->
                </div>
              </button></li>
          </ul>
        </div>
      </div>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
